server/dynamodb/events.go

```diff
@@ -282,10 +276,14 @@ func (db *DynamoDB) WriteContext(ctx context.Context, e *events.Event) error {
 	}
 
 	_, err = db.client.UpdateItemWithContext(ctx, item)
-	return err
+	return checkErrNotFound(err)
 }
 
-// Close closes the writer.
-func (db *DynamoDB) Close() error {
-	return nil
+func checkErrNotFound(err error) error {
+	if e, ok := err.(awserr.RequestFailure); ok {
+		if e.StatusCode() == 400 {
```

Comment:
This is too generic. 400 can mean lots of things. 

http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.MessagesAndCodes

Reply:
Hrm, I guess the best I can do is check for "ValidationException" then?

Reply:
I think you need to change how state change events are handled here.  Probably would be better to do something more similar to mongo or elastic here. 

I think that these updates will create an item in the database if it doesn't exist (http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateItem.html)

Reply:
we should probably have tests to see what happens in these event handlers when that task does not exist

Reply:
I agree. It was just something I realized thinking about this code. 

The main difference here is the error handling. I think the error handling should be reverted back to how it was before. 

The state transition code need to change to be able to handle canceling an unknown task.  I think this could be achieved in nearly identical manner to how it is implemented in mongo and elastic.

Reply:
Do you mean that the state change should use `GetTask` plus `tes.ValidateTransition` in order to be more like mongo and elastic?

As I read it, the `attemptItem` and `indexItem` updates here will fail with a `ValidationError` if the task doesn't exist, which is similar to mongo's and elastic's `ErrNotFound`. Failing if the task doesn't exist and returning `tes.ErrNotFound` seems like the right behavior, no?

Reply:
Or, maybe you mean that all updates should have an added conditional check that the task exists?

Reply:
> Do you mean that the state change should use GetTask plus tes.ValidateTransition in order to be more like mongo and elastic?

Yes.

> As I read it, the attemptItem and indexItem updates here will fail with a ValidationError if the task doesn't exist, which is similar to mongo's and elastic's ErrNotFound. Failing if the task doesn't exist and returning tes.ErrNotFound seems like the right behavior, no?

I don't think they will fail in that case. 

> Or, maybe you mean that all updates should have an added conditional check that the task exists?

Yeah all of these updates may need something along the lines of `ConditionExpression: "attribute_exists(<taskId>)` so that they will fail if the task doesn't exist.